File: src-tauri/src/colors.rs
//! colors.rs — Shared progress color mapping.
//!
//! Converts a progress ratio (0.0–1.0+) to colors used by the tray icon,
//! overlay bar, and UI components.

/// 0–60 % of session limit → neutral (below warning threshold).
pub const THRESHOLD_YELLOW: f32 = 0.60;
/// 60–85 % → yellow.
pub const THRESHOLD_RED: f32 = 0.85;

/// Neutral color for sitting below the warning threshold.
pub const NEUTRAL: (u8, u8, u8) = (0x2c, 0x29, 0x20);

/// Neutral dark base for completed standing laps (was goldenrod #B8860B).
pub const STANDING_LAP_BASE: (u8, u8, u8) = (0x24, 0x20, 0x18);

/// Returns (r, g, b) for a standing progress ratio.
///
/// Previously returned a gold gradient; now returns neutral to avoid green/gold signals.
pub fn color_for_standing(_progress: f32) -> (u8, u8, u8) {
    NEUTRAL
}

/// Returns (r, g, b, css_hex) for a given progress ratio.
///
/// Thresholds:
/// - < 60%: neutral (#2c2920)
/// - 60–85%: yellow (#ffc107)
/// - ≥ 85%: red (#f44336)
pub fn color_for_progress(progress: f32) -> (u8, u8, u8, &'static str) {
    if progress < THRESHOLD_YELLOW {
        (NEUTRAL.0, NEUTRAL.1, NEUTRAL.2, "#2c2920") // neutral
    } else if progress < THRESHOLD_RED {
        (255, 193, 7, "#ffc107") // yellow
    } else {
        (244, 67, 54, "#f44336") // red
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn color_neutral_below_60_percent() {
        let (r, g, b, css) = color_for_progress(0.3);
        assert_eq!((r, g, b), (0x2c, 0x29, 0x20));
        assert_eq!(css, "#2c2920");
    }

    #[test]
    fn color_yellow_between_60_and_85() {
        let (r, g, b, css) = color_for_progress(0.7);
        assert_eq!((r, g, b), (255, 193, 7));
        assert_eq!(css, "#ffc107");
    }

    #[test]
    fn color_red_above_85_percent() {
        let (r, g, b, css) = color_for_progress(0.9);
        assert_eq!((r, g, b), (244, 67, 54));
        assert_eq!(css, "#f44336");
    }

    #[test]
    fn color_at_zero_is_neutral() {
        let (r, g, b, css) = color_for_progress(0.0);
        assert_eq!((r, g, b), NEUTRAL);
        assert_eq!(css, "#2c2920");
    }

    #[test]
    fn color_at_60_percent_is_yellow() {
        let (_, _, _, css) = color_for_progress(0.60);
        assert_eq!(css, "#ffc107");
    }
}

File: src/utils/colors.ts
/**
 * colors.ts — Canonical color palette for desk session progress.
 *
 * Source of truth: `src-tauri/src/colors.rs`.
 * Both Rust (overlay bar) and TypeScript (popup ProgressBar, widgets) must
 * use the SAME hex values and thresholds defined here.
 *
 * If you change colors here, update `colors.rs` to match.
 */

/** Progress ratio below which the bar is neutral (no warning). */
export const THRESHOLD_YELLOW = 0.60;

/** Progress ratio at or above which the bar turns red. */
export const THRESHOLD_RED = 0.85;

/** Neutral color for sitting below warning threshold — matches colors.rs NEUTRAL. */
export const NEUTRAL = "#2c2920";

/** Sitting session progress colors — matches colors.rs exactly. */
export const SITTING_YELLOW = "#ffc107";
export const SITTING_RED = "#f44336";

/** Neutral dark base for completed standing laps — matches colors.rs STANDING_LAP_BASE. */
export const STANDING_LAP_BASE = "#242018";

/** Away state color. */
export const AWAY_GRAY = "#808080";

/**
 * Returns the state dot / label color for a given desk state.
 *
 * - Sitting: progress-based (neutral/yellow/red)
 * - Standing: beam amber (handled by CSS --beam variable)
 * - Away/Walking: gray
 */
export function stateColor(state: string, limitRatio = 0): string {
  if (state === "Sitting") return sittingColorForRatio(limitRatio);
  if (state === "Standing") return AWAY_GRAY;
  return AWAY_GRAY;
}

/**
 * Returns the sitting progress bar color for a given ratio (0.0-1.0).
 *
 * Thresholds (same as Rust `color_for_progress`):
 * - < 60%: neutral (#2c2920)
 * - 60-85%: yellow (#ffc107)
 * - >= 85%: red (#f44336)
 */
export function sittingColorForRatio(ratio: number): string {
  if (ratio >= THRESHOLD_RED) return SITTING_RED;
  if (ratio >= THRESHOLD_YELLOW) return SITTING_YELLOW;
  return NEUTRAL;
}
